start_time_intent_handler.py

```diff
@@ -57,88 +60,97 @@ def HandleIntent(self, post_params):
-                           ' You can also try asking prayer times for a city next time.')}
-
-      # the user has specified a state or country but not a city, then we should
-      # instruct them to tell us a city name.
-      if state or country:
-        return {'speech': ('Sorry, I don\'t have enough information. Please try '
-                           'again with a city next time.')}
-      
-      # at this stage, we don't know anything about the user's location - try
-      # checking our db for a cached location
-
-      # however, do not use the user's cached location if user has explicitly requested
-      # that their current location be used
-      explicit_location_requested = 'user-current-location' in params
-
-      user_info = self.fake_db_.GetUserInfo(user_id)
-      print 'user info for ', user_id, ' is ', user_info
-      if (not explicit_location_requested and user_info and 
-          user_info.get('lat') and user_info.get('lng') and user_info.get('city')):
-        print 'found user ', user_id, ' in databse, so location request is not necesary'
-        lat = user_info.get('lat')
-        lng = user_info.get('lng')
-        city = user_info.get('city')
-        return self._ComputePrayerTimeAndRespond(desired_prayer, lat, lng, city)
-
-      else:
-        print ('Could not find location in request, '
-               'so responding with a permission request.')
-        return response_builder.RequestLocationPermission()
-
-    # we must fill these parameters in order to make a query to salah.com
-    lat = None
-    lng = None
-
-    # if we have a city, then use this
-    if city:
-      print 'city:', city 
-      print 'country:', country 
-      print 'state:', state 
-      city = self._EncodeParameter(params.get('geo-city'))
-
-      if country:
-        country = self._EncodeParameter(params.get('geo-country'))
-
-      if state:
-        state = self._EncodeParameter(params.get('geo-state-us'))
-
-      location_coordinates = gmaps_API.GetGeocode(city, country, state)
-      lat = location_coordinates.get('lat')
-      lng = location_coordinates.get('lng')
-
-    # if we have a device location, then use it
-    elif has_location:
-      if permission_context:
-        print 'permission context'
-        location = device_params.get('location')
-        lat = location.get('coordinates').get('latitude')
-        lng = location.get('coordinates').get('longitude')
-        city = location.get('city')
-        if not city:
-          city = gmaps_API.LatLngToCity(lat, lng)
-        user_info = {'city': city, 'lat': lat, 'lng': lng}
-        print 'caching user location for ', user_id, ' as ', json.dumps(user_info)
-        self.fake_db_.AddOrUpdateUser(user_id, user_info)
-
-      else:
-        print 'Could not find relevant context!'
-
-    return self._ComputePrayerTimeAndRespond(desired_prayer, lat, lng, city)
+    # if there is no masjid then we need to pass the city / location to salah.com
+    if not (masjid):
```

Comment:
this is starting to nest way too much.. there's an alternate way to structure this code such that if we see masjid, then call some helper method to handle masjid requests - since masjid requests will always be responded to differently than those based on location.

can we try:

if masjid:
    return HandleMasjidRequest

... leave rest of logic untouched below